File: pkg/storage/resource.go
package storage

import (
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"strings"
	"sxwl/3k/pkg/consts"
)

func ResourceToOSSPath(resourceType, resource string) string {
	switch resourceType {
	case consts.Model:
		if strings.HasPrefix(resource, "user-") {
			return fmt.Sprintf(consts.OSSUserModelPath, resource)
		} else {
			return fmt.Sprintf(consts.OSSPublicModelPath, resource)
		}
	case consts.Dataset:
		if strings.HasPrefix(resource, "user-") {
			return fmt.Sprintf(consts.OSSUserDatasetPath, resource)
		} else {
			return fmt.Sprintf(consts.OSSPublicDatasetPath, resource)
		}
	case consts.Adapter:
		if strings.HasPrefix(resource, "user-") {
			return fmt.Sprintf(consts.OSSUserAdapterPath, resource)
		} else {
			return fmt.Sprintf(consts.OSSPublicAdapterPath, resource)
		}
	}

	return ""
}

func ResourceIsPublic(resource string) bool {
	if strings.HasPrefix(resource, "user-") {
		return false
	} else {
		return true
	}
}

func ModelCRDName(ossPath string) string {
	return fmt.Sprintf("model-storage-%s", hash(ossPath))
}

func DatasetCRDName(ossPath string) string {
	return fmt.Sprintf("dataset-storage-%s", hash(ossPath))
}

func AdapterCRDName(ossPath string) string {
	return fmt.Sprintf("adapter-storage-%s", hash(ossPath))
}

func ModelPVCName(ossPath string) string {
	return fmt.Sprintf("pvc-model-%s", hash(ossPath))
}

func DatasetPVCName(ossPath string) string {
	return fmt.Sprintf("pvc-dataset-%s", hash(ossPath))
}

func AdapterPVCName(ossPath string) string {
	return fmt.Sprintf("pvc-adapter-%s", hash(ossPath))
}

func ModelDownloadJobName(ossPath string) string {
	return fmt.Sprintf("download-model-%s", hash(ossPath))
}

func DatasetDownloadJobName(ossPath string) string {
	return fmt.Sprintf("download-dataset-%s", hash(ossPath))
}

func AdapterDownloadJobName(ossPath string) string {
	return fmt.Sprintf("download-adapter-%s", hash(ossPath))
}

func OssPathToOssURL(bucket, ossPath string) string {
	return fmt.Sprintf("oss://%s/%s", bucket, ossPath)
}

func hash(data string) string {
	hash := sha1.New()
	hash.Write([]byte(data))
	hashed := hash.Sum(nil)
	return hex.EncodeToString(hashed)[:16]
}

func ExtractTemplate(filename string) string {
	// Split the filename by "/"
	parts := strings.Split(filename, "/")
	// Get the last part of the split, which should be "<template>-<name>.<extension>"
	lastPart := parts[len(parts)-1]
	// Now split the last part by "-" and get the second to last element, which is assumed to be the template name
	nameParts := strings.Split(lastPart, "-")
	// Before extracting the name, we split the last part of nameParts by "." to remove the file extension
	templateWithExtension := nameParts[len(nameParts)-1]
	template := strings.Split(templateWithExtension, ".")[0]
	return template
}

File: 3kctl/download/resource.py
import hashlib

# Placeholder constants - replace these with your actual values
MODEL = "model"
DATASET = "dataset"
OSS_USER_MODEL_PATH = "models/{}"
OSS_PUBLIC_MODEL_PATH = "models/public/{}"
OSS_USER_DATASET_PATH = "datasets/{}"
OSS_PUBLIC_DATASET_PATH = "datasets/public/{}"


def resource_to_oss_path(resource_type: str, resource: str) -> str:
    if resource_type == MODEL:
        if resource.startswith("user-"):
            return OSS_USER_MODEL_PATH.format(resource)
        else:
            return OSS_PUBLIC_MODEL_PATH.format(resource)
    elif resource_type == DATASET:
        if resource.startswith("user-"):
            return OSS_USER_DATASET_PATH.format(resource)
        else:
            return OSS_PUBLIC_DATASET_PATH.format(resource)

    return ""


def model_crd_name(oss_path: str) -> str:
    return f"model-storage-{hash_data(oss_path)}"


def dataset_crd_name(oss_path: str) -> str:
    return f"dataset-storage-{hash_data(oss_path)}"


def model_pvc_name(oss_path: str) -> str:
    return f"pvc-model-{hash_data(oss_path)}"


def dataset_pvc_name(oss_path: str) -> str:
    return f"pvc-dataset-{hash_data(oss_path)}"


def model_download_job_name(oss_path: str) -> str:
    return f"download-model-{hash_data(oss_path)}"


def dataset_download_job_name(oss_path: str) -> str:
    return f"download-dataset-{hash_data(oss_path)}"


def oss_path_to_oss_url(bucket: str, oss_path: str) -> str:
    return f"oss://{bucket}/{oss_path}"


def hash_data(data: str) -> str:
    hasher = hashlib.sha1()
    hasher.update(data.encode('utf-8'))
    hashed = hasher.hexdigest()
    return hashed[:16]
